designs/node-repair.md

```diff
@@ -0,0 +1,95 @@
+Main drawback of this option is providers will need to implement much more undifferentiated logic, without having the ability to model duration and remediation on nodes. The reconciliation step for replacing an unhealthy node will look like: 
+
+1. Karpenter will reconcile on the NodeClaim and make a cloud provide call with IsHealthy
+2. If the node is determined to unhealthy, a status condition will be added to the NodeClaim
+3. Disruption controller will force terminate the node claim immediately 
+
+### Option 3: Cloud provider injected unhealthy condition 
+
+```
+type UnhealthyType metav1.ConditionType 
+type UnhealthyRepairAfter time.Duration
+
+// UnhealthyConditions are condition to watch for on nodes inside of the cluster
+var UnhealthyConditions = map[UnhealthyType]UnhealthyRepairAfter{
+    "Ready": "30m"
+    "DiskPressure": "10m"
+    ...
+}
+```
+
+Cloud provider hydrate a `UnhealthyConditions` map that will hold the condition to track on nodes. The map will track the condition and `UnhealthyRepairAfter` will dictate the length of time the node will be inside the cluster before action is taken against it. Node health validation will follow as such:
+
+1. A diagnostic agent will add a status condition on to a node 
+2. Karpenter will reconcile on nodes and identify unhealthy nodes defined by `UnhealthyConditions` map
+3. Unhealthy condition is added to the NodeClaim
+4. Disruption controller will forcefully terminate the the NodeClaim once the node has been in an unhealthy state for the duration specified by `UnhealthyConditions` map
+
+## Recommended Solution
+
+The recommended approach is for cloud provider interface should expose a layer of configuration for Karpenter to monitor nodes and take action. I propose we implement option one. Cloud provider can define any set of status conditions that are tracked on the node. An interface approach allows for easy extensibility. Option 2 has the approach of being over generalized as we don’t have use-cases to show the need. All solutions outlined below will be under a feature flag to indicate that these feature are not stable. 
+
+### Forceful termination
+
+For a first iteration approach Karpenter will implement force termination. Today, graceful termination in Karpenter will attempt to wait for pod to be fully drained on a node and all volume attachment to be deleted from a node. This raises the problem that during graceful termination node can be stuck terminating as pod eviction or volume detachment may be broken. In these cases, users will need to take manual action against the node or if set, termination grace period will force terminate the node. For the first iteration of problem, the recommendation will be to force terminate nodes now and follow-up to allow configurability for either the cloud providers or the Karpenter users once we have sufficient data to support the need for graceful termination for unhealthy nodes. 
```

Comment:
Thoughts on having other autohealing actions that we can take? Repair Policy allows us to define conditions for unhealthy, but also it might make sense to have an RepairAction the customer can eventually configure. 

AKS has found great success to repairing node not ready by simply issuing a vm.Restart operation to azure, there are some benefits to not fully deleting the vm and instead doing softer operations for repair. 

This is out of scope for this design but wanted to mention it so its in peoples minds.

Reply:
https://learn.microsoft.com/en-us/azure/aks/node-auto-repair for reference/inspiration

Reply:
https://cloud.google.com/kubernetes-engine/docs/how-to/node-auto-repair